src/jit/lowerxarch.cpp

```diff
@@ -3869,27 +3897,103 @@ bool Lowering:: IsContainableImmed(GenTree* parentNode, GenTree* childNode)
     return true;
 }
 
-//----------------------------------------------------------------------
-// TryToSetRegOptional - sets a bit to indicate to LSRA that register
-// for a given tree node is optional for codegen purpose.  If no
-// register is allocated to such a tree node, its parent node treats
-// it as a contained memory operand during codegen.
+//-----------------------------------------------------------------------
+// IsProfitableToMarkOp1AsRegOptional: returns true if it is likely to 
+// be profitable to mark op1 of a binary-op tree as reg optional.
+//
+// Since only one of op1 or op2 can be a memory operand on xarch, only
+// one of  them have to be marked as reg optional.  Since Lower doesn't
+// know apriori which of op1 or op2 is not likely to get a register, it
+// has to make a guess. This routine encapsulates heuristics that
+// guess whether it is likely to be beneficial to mark op1 as reg
+// optional.  If this routine returns false, then it is beneficial to
+// mark op2 as reg optional.
+// 
 //
 // Arguments:
-//    tree    -   GenTree node
+//     tree  -  a binary-op tree node
 //
-// Returns
-//    None
+// Returns:
+//     true if it is likely to be beneficial marking op1 as reg optional.
+//     false if it is beneficial to mark op2 as reg optional.
 //
-// Note: Right now a tree node is marked as reg optional only
-// if is it a GT_LCL_VAR.  This routine needs to be modified if
-// in future if lower/codegen needs to support other tree node
-// types.
-void Lowering::TryToSetRegOptional(GenTree* tree)
+bool Lowering::IsProfitableToMarkOp1AsRegOptional(GenTree* tree)
 {
-    if (tree->OperGet() == GT_LCL_VAR)
+    assert(GenTree::OperIsBinary(tree->OperGet()));
+    assert(tree->OperIsCommutative() || tree->OperIsCompare());
+
+    GenTree* op1 = tree->gtGetOp1();
+    GenTree* op2 = tree->gtGetOp2();
+
+    // This routine uses the following heuristics:
+    //
+    // a) If both are tracked locals, marking the one with lower weighted
+    // ref count as reg-optional would likely be beneficial as it has
+    // higher probability of not getting a register.
+    //
+    // b) op1 = tracked local and op2 = untracked local: op2's def position
+    // requires a reg and hence it is beneficial to mark op1 as reg optional.
```

Comment:
It should not be the case that an untracked local requires a register. Or is it the case that this will not be called, because if it could be RegOptional it would already be contained? In any case, that should be clarified